File: ash-zig/src/safety/guards.zig
const std = @import("std");
const utils = @import("../utils.zig");

/// Paths that must never be deleted
const never_delete = [_][]const u8{
    "~/Library/Keychains",
    "~/.ssh",
    "~/.gnupg",
    "~/.config",
    "~/.local/share",
    "/System",
    "/usr",
    "/bin",
    "/sbin",
    "/private/var/vm",
    "/private/var/db",
    "/Applications",
    "/Library/Keychains",
    "/Network",
    "/cores",
};

/// File patterns that must never be deleted
const never_delete_patterns = [_][]const u8{
    "*.keychain*",
    "*.pem",
    "*.key",
    ".git",
    ".gitignore",
    "id_rsa*",
    "id_ed25519*",
    "*.p12",
    "*.pfx",
    "*.cer",
    "*.crt",
    "known_hosts",
    "authorized_keys",
};

/// Bundle ID prefixes for protected apps
const protected_bundle_ids = [_][]const u8{
    "com.apple.",
    "com.microsoft.",
};

/// Size threshold (1 GB) that requires confirmation
pub const size_confirmation_threshold: u64 = 1024 * 1024 * 1024;

/// Check if a path is safe to delete
pub fn isSafePath(allocator: std.mem.Allocator, path: []const u8) bool {
    // Expand the path
    const expanded = utils.expandPath(allocator, path) catch return false;
    defer allocator.free(expanded);

    // Check never-delete directories
    for (never_delete) |blocked| {
        const blocked_expanded = utils.expandPath(allocator, blocked) catch continue;
        defer allocator.free(blocked_expanded);

        if (utils.pathStartsWith(expanded, blocked_expanded)) {
            return false;
        }
    }

    // Check never-delete patterns
    const base = std.fs.path.basename(path);
    for (never_delete_patterns) |pattern| {
        if (utils.matchGlob(base, pattern)) {
            return false;
        }
    }

    // Check for .git anywhere in path
    if (std.mem.indexOf(u8, path, "/.git/") != null or
        std.mem.indexOf(u8, path, "/.git") != null)
    {
        return false;
    }

    return true;
}

/// Check if a bundle ID is protected
pub fn isProtectedApp(bundle_id: []const u8) bool {
    for (protected_bundle_ids) |prefix| {
        if (std.mem.startsWith(u8, bundle_id, prefix)) {
            return true;
        }
    }
    return false;
}

/// Check if a path requires confirmation before deletion
pub fn requiresConfirmation(allocator: std.mem.Allocator, path: []const u8, size: u64) bool {
    // Large files always require confirmation
    if (size >= size_confirmation_threshold) {
        return true;
    }

    const expanded = utils.expandPath(allocator, path) catch return true;
    defer allocator.free(expanded);

    // iOS backups
    if (std.mem.indexOf(u8, expanded, "MobileSync/Backup") != null) {
        return true;
    }

    // Xcode Archives
    if (std.mem.indexOf(u8, expanded, "Xcode/Archives") != null) {
        return true;
    }

    // Application Support directories
    if (std.mem.indexOf(u8, expanded, "Application Support") != null) {
        return true;
    }

    return false;
}

/// Validate a list of paths, returning (safe_paths, blocked_paths)
pub fn validatePaths(
    allocator: std.mem.Allocator,
    paths: []const []const u8,
) !struct { safe: std.ArrayList([]const u8), blocked: std.ArrayList([]const u8) } {
    var safe = std.ArrayList([]const u8).init(allocator);
    var blocked = std.ArrayList([]const u8).init(allocator);

    for (paths) |path| {
        if (isSafePath(allocator, path)) {
            try safe.append(path);
        } else {
            try blocked.append(path);
        }
    }

    return .{ .safe = safe, .blocked = blocked };
}

/// Sanitize a path by removing dangerous characters
pub fn sanitizePath(allocator: std.mem.Allocator, path: []const u8) ![]const u8 {
    var result = std.ArrayList(u8).init(allocator);
    errdefer result.deinit();

    for (path) |c| {
        // Skip null bytes and other control characters
        if (c == 0 or (c < 32 and c != '\t')) continue;
        try result.append(c);
    }

    return result.toOwnedSlice();
}

// Tests
test "isSafePath" {
    const allocator = std.testing.allocator;

    // Safe paths
    try std.testing.expect(isSafePath(allocator, "~/Library/Caches/com.example.app"));
    try std.testing.expect(isSafePath(allocator, "~/Library/Logs/app.log"));
    try std.testing.expect(isSafePath(allocator, "~/Library/Developer/Xcode/DerivedData/Project"));

    // Dangerous paths - must return false
    try std.testing.expect(!isSafePath(allocator, "~/Library/Keychains/login.keychain-db"));
    try std.testing.expect(!isSafePath(allocator, "~/.ssh/id_ed25519"));
    try std.testing.expect(!isSafePath(allocator, "~/.gnupg/private-keys"));
    try std.testing.expect(!isSafePath(allocator, "/System/Library/CoreServices"));
    try std.testing.expect(!isSafePath(allocator, "/usr/bin/bash"));
    try std.testing.expect(!isSafePath(allocator, "~/projects/app/.git/objects"));
}

test "isProtectedApp" {
    try std.testing.expect(isProtectedApp("com.apple.Safari"));
    try std.testing.expect(isProtectedApp("com.apple.finder"));
    try std.testing.expect(isProtectedApp("com.microsoft.VSCode"));
    try std.testing.expect(!isProtectedApp("com.spotify.client"));
    try std.testing.expect(!isProtectedApp("com.homebrew.cask"));
}

test "requiresConfirmation" {
    const allocator = std.testing.allocator;

    // Large files
    try std.testing.expect(requiresConfirmation(allocator, "/tmp/large", 2 * 1024 * 1024 * 1024));

    // Small files
    try std.testing.expect(!requiresConfirmation(allocator, "/tmp/small", 1024));
}

File: internal/safety/guards.go
package safety

import (
	"os"
	"path/filepath"
	"strings"
)

// neverDelete contains paths that should never be deleted.
var neverDelete = []string{
	"~/Library/Keychains",
	"~/.ssh",
	"~/.gnupg",
	"~/.config",
	"~/.local/share",
	"/System",
	"/usr",
	"/bin",
	"/sbin",
	"/private/var/vm",
	"/private/var/db",
	"/Applications",
	"/Library/Keychains",
	"/Network",
	"/cores",
}

// neverDeletePatterns contains file patterns that should never be deleted.
var neverDeletePatterns = []string{
	"*.keychain*",
	"*.pem",
	"*.key",
	".git",
	".gitignore",
	"id_rsa*",
	"id_ed25519*",
	"*.p12",
	"*.pfx",
	"*.cer",
	"*.crt",
	"known_hosts",
	"authorized_keys",
}

// protectedBundleIDs contains bundle ID prefixes that should be protected.
var protectedBundleIDs = []string{
	"com.apple.",
	"com.microsoft.",
}

// SizeConfirmationThreshold is the size (1GB) above which items require confirmation.
const SizeConfirmationThreshold int64 = 1024 * 1024 * 1024

// IsSafePath checks if a path is safe to delete.
func IsSafePath(path string) bool {
	expanded := filepath.Clean(expandPath(path))
	originalExpanded := expanded

	if isBlockedSymlinkTarget(originalExpanded) {
		return false
	}

	// Resolve symlinks to prevent bypass attacks (e.g., symlink pointing to ~/.ssh)
	resolved, err := filepath.EvalSymlinks(expanded)
	if err == nil {
		expanded = resolved
	}
	// If symlink resolution fails (e.g., broken symlink), continue with original path

	if isBlockedPath(expanded, originalExpanded) || matchesNeverDeletePattern(path) || containsGitDir(path) {
		return false
	}

	return true
}

// IsProtectedApp checks if a bundle ID belongs to a protected app.
func IsProtectedApp(bundleID string) bool {
	for _, prefix := range protectedBundleIDs {
		if strings.HasPrefix(bundleID, prefix) {
			return true
		}
	}
	return false
}

// RequiresConfirmation checks if a path requires explicit user confirmation.
func RequiresConfirmation(path string, size int64) bool {
	expanded := expandPath(path)

	// iOS backups
	if strings.Contains(expanded, "MobileSync/Backup") {
		return true
	}

	// Xcode Archives
	if strings.Contains(expanded, "Xcode/Archives") {
		return true
	}

	// Large items (> 1GB)
	if size > SizeConfirmationThreshold {
		return true
	}

	// Application Support
	if strings.Contains(expanded, "Application Support") {
		return true
	}

	return false
}

// GetNeverDeletePaths returns a copy of the paths that should never be deleted.
func GetNeverDeletePaths() []string {
	result := make([]string, len(neverDelete))
	copy(result, neverDelete)
	return result
}

// GetNeverDeletePatterns returns a copy of the patterns that should never be deleted.
func GetNeverDeletePatterns() []string {
	result := make([]string, len(neverDeletePatterns))
	copy(result, neverDeletePatterns)
	return result
}

// GetProtectedBundleIDs returns a copy of the protected bundle ID prefixes.
func GetProtectedBundleIDs() []string {
	result := make([]string, len(protectedBundleIDs))
	copy(result, protectedBundleIDs)
	return result
}

func expandPath(path string) string {
	if strings.HasPrefix(path, "~/") {
		home, err := os.UserHomeDir()
		if err != nil {
			return path
		}
		return filepath.Join(home, path[2:])
	}
	return path
}

func containsGitDir(path string) bool {
	parts := strings.Split(filepath.ToSlash(path), "/")
	for _, part := range parts {
		if part == ".git" {
			return true
		}
	}
	return false
}

func pathHasPrefix(path, prefix string) bool {
	rel, err := filepath.Rel(prefix, path)
	if err != nil {
		return false
	}
	return rel == "." || (!strings.HasPrefix(rel, ".."+string(filepath.Separator)) && rel != "..")
}

func readSymlinkTarget(path string) (string, bool) {
	info, err := os.Lstat(path)
	if err != nil || info.Mode()&os.ModeSymlink == 0 {
		return "", false
	}

	target, err := os.Readlink(path)
	if err != nil {
		return "", false
	}
	if filepath.IsAbs(target) {
		return target, true
	}
	return filepath.Join(filepath.Dir(path), target), true
}

func isBlockedSymlinkTarget(path string) bool {
	symlinkTarget, ok := readSymlinkTarget(path)
	if !ok {
		return false
	}

	expandedTarget := filepath.Clean(expandPath(symlinkTarget))
	for _, blocked := range neverDelete {
		if pathHasPrefix(expandedTarget, filepath.Clean(expandPath(blocked))) {
			return true
		}
	}

	return false
}

func isBlockedPath(resolvedPath, originalPath string) bool {
	for _, blocked := range neverDelete {
		blockedExpanded := filepath.Clean(expandPath(blocked))
		if pathHasPrefix(resolvedPath, blockedExpanded) || pathHasPrefix(originalPath, blockedExpanded) {
			return true
		}
	}

	return false
}

func matchesNeverDeletePattern(path string) bool {
	base := filepath.Base(path)
	for _, pattern := range neverDeletePatterns {
		matched, err := filepath.Match(pattern, base)
		if err != nil || matched {
			return true
		}
	}

	return false
}

// ValidatePaths checks a list of paths and returns only safe ones.
func ValidatePaths(paths []string) (safe []string, blocked []string) {
	for _, path := range paths {
		if IsSafePath(path) {
			safe = append(safe, path)
		} else {
			blocked = append(blocked, path)
		}
	}
	return safe, blocked
}

// SanitizePath removes any potentially dangerous characters from a path.
func SanitizePath(path string) string {
	// Remove null bytes and other control characters
	var sanitized strings.Builder
	for _, r := range path {
		if r >= 32 && r != 127 {
			sanitized.WriteRune(r)
		}
	}
	return sanitized.String()
}
